google-cloud-spanner-hibernate-dialect/src/main/java/com/google/cloud/spanner/hibernate/types/SpannerJsonType.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2019-2020 Google LLC
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
+ */
+
+package com.google.cloud.spanner.hibernate.types;
+
+import com.google.cloud.spanner.hibernate.types.internal.JsonJavaTypeDescriptor;
+import com.google.cloud.spanner.hibernate.types.internal.JsonSqlTypeDescriptor;
+import java.util.Properties;
+import org.hibernate.type.AbstractSingleColumnStandardBasicType;
+import org.hibernate.usertype.DynamicParameterizedType;
+
+/**
+ * A {@link org.hibernate.usertype.UserType} which supports mapping Spanner JSON columns to
+ * entity fields.
+ */
+public class SpannerJsonType extends AbstractSingleColumnStandardBasicType<Object>
+    implements DynamicParameterizedType {
+
+  public SpannerJsonType() {
+    super(new JsonSqlTypeDescriptor(), new JsonJavaTypeDescriptor());
```

Comment:
Actually I can't quite do this. So a `SpannerJsonType` gets created for each field on an entity that receives the JSON type annotation. However, this makes it so we cannot make `JsonJavaTypeDescriptor` static, since it carries state about what was the class type in the entity that received the annotation: https://github.com/GoogleCloudPlatform/google-cloud-spanner-hibernate/pull/321/files#diff-098729681e5bbcfe85a1db02a5402cccf0824e4683deed0e1231e65d6a385d8d
And it is this type that we convert from JSON string to object with using GSON.

I could theoretically make the `JsonSqlTypeDescriptor` static though since it does not carry state, but I think it would just be confusing to have one static and one not static, and lead to bugs if it needs to carry state in the future.

Hibernate-types also makes separate instances: https://github.com/vladmihalcea/hibernate-types/blob/master/hibernate-types-55/src/main/java/com/vladmihalcea/hibernate/type/json/JsonType.java#L42

Reply:
Fair enough, I'm not sure where I saw them make these static.